In [1]:
import numpy as np
import pandas as pd

In [2]:
# numpy version check
print(np.__version__)
# Pandas version
print(pd.__version__)

1.15.4
0.23.4


In [3]:
# Read the data
data = pd.read_csv("FL_insurance_sample.csv")

In [4]:
# Checking output type
print(type(data))

<class 'pandas.core.frame.DataFrame'>


In [5]:
# to check the shape
print(data.shape)
# to get the number of rows or instances or records or examples
print(data.shape[0])
# to get the number of columns or features or attributes
print(data.shape[1])

(13571, 18)
13571
18


In [6]:
# To get the column names
print(data.columns)
# To get the row names
print(data.index)

Index(['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity'],
      dtype='object')
RangeIndex(start=0, stop=13571, step=1)


In [7]:
# Datatype of each column
print(data.dtypes)
print("-----"*15)
# what is output type for data.dtypes
print(type(data.dtypes))
print("-----"*15)
# Access the first column dtype
print(data.dtypes[0])
print("-----"*15)
# Access the last column dtype
print(data.dtypes[-1])

policyID                int64
statecode              object
county                 object
eq_site_limit         float64
hu_site_limit         float64
fl_site_limit         float64
fr_site_limit         float64
tiv_2011              float64
tiv_2012              float64
eq_site_deductible    float64
hu_site_deductible    float64
fl_site_deductible    float64
fr_site_deductible      int64
point_latitude        float64
point_longitude       float64
line                   object
construction           object
point_granularity     float64
dtype: object
---------------------------------------------------------------------------
<class 'pandas.core.series.Series'>
---------------------------------------------------------------------------
int64
---------------------------------------------------------------------------
float64


In [8]:
# Column count for each datatype
# How many columns are there with int64
# How many columns are there with float64
# How many columns are there with object
data.dtypes.value_counts()
# Note We are using series method value_counts

float64    12
object      4
int64       2
dtype: int64

# observe the data

In [9]:
# display top 5 rows
data.head()

policyID statecode       county  eq_site_limit  hu_site_limit  \
0    119736        FL  CLAY COUNTY       498960.0      498960.00   
1    448094        FL  CLAY COUNTY      1322376.3     1322376.30   
2    206893        FL  CLAY COUNTY       190724.4      190724.40   
3    333743        FL  CLAY COUNTY            0.0       79520.76   
4    172534        FL  CLAY COUNTY            0.0      254281.50   

   fl_site_limit  fr_site_limit    tiv_2011    tiv_2012  eq_site_deductible  \
0       498960.0       498960.0   498960.00   792148.90                 0.0   
1      1322376.3      1322376.3  1322376.30  1438163.57                 0.0   
2       190724.4       190724.4   190724.40   192476.78                 0.0   
3            0.0            0.0    79520.76    86854.48                 0.0   
4            0.0       254281.5   254281.50   246144.49                 0.0   

   hu_site_deductible  fl_site_deductible  fr_site_deductible  point_latitude  \
0              9979.2                 0.0                   0       30.102261   
1                 0.0                 0.0                   0       30.063936   
2                 0.0                 0.0                   0       30.089579   
3                 0.0                 0.0                   0       30.063236   
4                 0.0                 0.0                   0       30.060614   

   point_longitude         line construction  point_granularity  
0       -81.711777  Residential      Masonry                1.0  
1       -81.707664  Residential      Masonry                3.0  
2       -81.700455  Residential         Wood                1.0  
3       -81.707703  Residential         Wood                3.0  
4       -81.702675  Residential         Wood                1.0

In [10]:
# display top 10 rows
data.head(10)

policyID statecode       county  eq_site_limit  hu_site_limit  \
0    119736        FL  CLAY COUNTY       498960.0      498960.00   
1    448094        FL  CLAY COUNTY      1322376.3     1322376.30   
2    206893        FL  CLAY COUNTY       190724.4      190724.40   
3    333743        FL  CLAY COUNTY            0.0       79520.76   
4    172534        FL  CLAY COUNTY            0.0      254281.50   
5    785275        FL  CLAY COUNTY            0.0      515035.62   
6    995932        FL  CLAY COUNTY            0.0    19260000.00   
7    223488        FL  CLAY COUNTY       328500.0      328500.00   
8    433512        FL  CLAY COUNTY       315000.0      315000.00   
9    142071        FL  CLAY COUNTY       705600.0      705600.00   

   fl_site_limit  fr_site_limit     tiv_2011     tiv_2012  eq_site_deductible  \
0       498960.0       498960.0    498960.00    792148.90                 0.0   
1      1322376.3      1322376.3   1322376.30   1438163.57                 0.0   
2       190724.4       190724.4    190724.40    192476.78                 0.0   
3            0.0            0.0     79520.76     86854.48                 0.0   
4            0.0       254281.5    254281.50    246144.49                 0.0   
5            0.0            0.0    515035.62    884419.17                 0.0   
6            0.0            0.0  19260000.00  20610000.00                 0.0   
7       328500.0       328500.0    328500.00    348374.25                 0.0   
8       315000.0       315000.0    315000.00    265821.57                 0.0   
9       705600.0       705600.0    705600.00   1010842.56             14112.0   

   hu_site_deductible  fl_site_deductible  fr_site_deductible  point_latitude  \
0              9979.2                 0.0                   0       30.102261   
1                 0.0                 0.0                   0       30.063936   
2                 0.0                 0.0                   0       30.089579   
3                 0.0                 0.0                   0       30.063236   
4                 0.0                 0.0                   0       30.060614   
5                 0.0                 0.0                   0       30.063236   
6                 0.0                 0.0                   0       30.102226   
7             16425.0                 0.0                   0       30.102217   
8             15750.0                 0.0                   0       30.118774   
9             35280.0                 0.0                   0       30.100628   

   point_longitude         line         construction  point_granularity  
0       -81.711777  Residential              Masonry                1.0  
1       -81.707664  Residential              Masonry                3.0  
2       -81.700455  Residential                 Wood                1.0  
3       -81.707703  Residential                 Wood                3.0  
4       -81.702675  Residential                 Wood                1.0  
5       -81.707703  Residential              Masonry                3.0  
6       -81.713882   Commercial  Reinforced Concrete                1.0  
7       -81.707146  Residential                 Wood                1.0  
8       -81.704613  Residential                 Wood                1.0  
9       -81.703751  Residential              Masonry                1.0

In [11]:
# display last 5 rows
data.tail()

policyID statecode         county  eq_site_limit  hu_site_limit  \
13566    906943        FL  ORANGE COUNTY            0.0     1566401.56   
13567    758044        FL  ORANGE COUNTY      4791330.0     4791330.00   
13568    387570        FL  ORANGE COUNTY      5133937.5     5133937.50   
13569    310787        FL  ORANGE COUNTY       918000.0      918000.00   
13570    522718        FL  ORANGE COUNTY      1349172.0     1349172.00   

       fl_site_limit  fr_site_limit    tiv_2011    tiv_2012  \
13566            0.0            0.0  1566401.56  1321002.82   
13567      4791330.0      4791330.0  4791330.00  7512163.40   
13568      5133937.5      5133937.5  5133937.50  8120605.64   
13569       918000.0       918000.0   918000.00   875191.82   
13570      1349172.0      1349172.0  1349172.00  2115013.30   

       eq_site_deductible  hu_site_deductible  fl_site_deductible  \
13566                 0.0                 0.0                 0.0   
13567                 0.0                 0.0                 0.0   
13568                 0.0                 0.0                 0.0   
13569                 0.0                 0.0                 0.0   
13570                 0.0                 0.0                 0.0   

       fr_site_deductible  point_latitude  point_longitude         line  \
13566                   0       28.456340       -81.470120  Residential   
13567                   0       28.468555       -81.450493   Commercial   
13568                   0       28.443203       -81.421188   Commercial   
13569                   0       28.442978       -81.421326  Residential   
13570                   0       28.442978       -81.421000          NaN   

             construction  point_granularity  
13566             Masonry                1.0  
13567  Reinforced Masonry                1.0  
13568  Reinforced Masonry                3.0  
13569             Masonry                3.0  
13570                 NaN                NaN

In [12]:

# display last 3 rows
data.tail(3)

policyID statecode         county  eq_site_limit  hu_site_limit  \
13568    387570        FL  ORANGE COUNTY      5133937.5      5133937.5   
13569    310787        FL  ORANGE COUNTY       918000.0       918000.0   
13570    522718        FL  ORANGE COUNTY      1349172.0      1349172.0   

       fl_site_limit  fr_site_limit   tiv_2011    tiv_2012  \
13568      5133937.5      5133937.5  5133937.5  8120605.64   
13569       918000.0       918000.0   918000.0   875191.82   
13570      1349172.0      1349172.0  1349172.0  2115013.30   

       eq_site_deductible  hu_site_deductible  fl_site_deductible  \
13568                 0.0                 0.0                 0.0   
13569                 0.0                 0.0                 0.0   
13570                 0.0                 0.0                 0.0   

       fr_site_deductible  point_latitude  point_longitude         line  \
13568                   0       28.443203       -81.421188   Commercial   
13569                   0       28.442978       -81.421326  Residential   
13570                   0       28.442978       -81.421000          NaN   

             construction  point_granularity  
13568  Reinforced Masonry                3.0  
13569             Masonry                3.0  
13570                 NaN                NaN

In [13]:
# select a random row
data.sample()

policyID statecode             county  eq_site_limit  hu_site_limit  \
8914    607816        FL  SANTA ROSA COUNTY        40500.0        40500.0   

      fl_site_limit  fr_site_limit  tiv_2011  tiv_2012  eq_site_deductible  \
8914        40500.0        40500.0   40500.0   38321.1                 0.0   

      hu_site_deductible  fl_site_deductible  fr_site_deductible  \
8914               810.0                 0.0                   0   

      point_latitude  point_longitude         line construction  \
8914       30.603575       -87.153015  Residential         Wood   

      point_granularity  
8914                1.0

In [14]:
# select 5 rows randomly
data.sample(5)

policyID statecode           county  eq_site_limit  hu_site_limit  \
11545    101438        FL  SEMINOLE COUNTY            0.0      590117.65   
4495     186330        FL     DUVAL COUNTY      1741945.5     1741945.50   
13379    541334        FL    ORANGE COUNTY            0.0      216000.00   
4011     698079        FL     DUVAL COUNTY      1073771.1     1073771.10   
5520     954047        FL    TAYLOR COUNTY         7542.0        7542.00   

       fl_site_limit  fr_site_limit    tiv_2011    tiv_2012  \
11545            0.0            0.0   590117.65   808402.16   
4495       1741945.5      1741945.5  1741945.50  2847356.24   
13379            0.0       216000.0   216000.00   332563.97   
4011             0.0            0.0  1073771.10  1135351.87   
5520          7542.0         7542.0     7542.00    10356.64   

       eq_site_deductible  hu_site_deductible  fl_site_deductible  \
11545                 0.0                 0.0                 0.0   
4495                  0.0                 0.0                 0.0   
13379                 0.0              6480.0                 0.0   
4011                  0.0                 0.0                 0.0   
5520                  0.0                 0.0                 0.0   

       fr_site_deductible  point_latitude  point_longitude         line  \
11545                   0       28.649985       -81.329980  Residential   
4495                    0       30.241779       -81.598534  Residential   
13379                   0       28.546460       -81.439430  Residential   
4011                    0       30.322203       -81.442009  Residential   
5520                    0       30.123390       -83.570140  Residential   

      construction  point_granularity  
11545      Masonry                3.0  
4495       Masonry                1.0  
13379         Wood                1.0  
4011       Masonry                3.0  
5520          Wood                1.0

- data.head()
- data.tail()
- data.samle()

 the output type of all the above methods are dataframe

# Info about the dataframe

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13571 entries, 0 to 13570
Data columns (total 18 columns):
policyID              13571 non-null int64
statecode             13571 non-null object
county                13571 non-null object
eq_site_limit         13571 non-null float64
hu_site_limit         13571 non-null float64
fl_site_limit         13571 non-null float64
fr_site_limit         13571 non-null float64
tiv_2011              13571 non-null float64
tiv_2012              13571 non-null float64
eq_site_deductible    13571 non-null float64
hu_site_deductible    13571 non-null float64
fl_site_deductible    13571 non-null float64
fr_site_deductible    13571 non-null int64
point_latitude        13571 non-null float64
point_longitude       13571 non-null float64
line                  13570 non-null object
construction          13570 non-null object
point_granularity     13570 non-null float64
dtypes: float64(12), int64(2), object(4)
memory usage: 1.9+ MB


In [16]:
type(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13571 entries, 0 to 13570
Data columns (total 18 columns):
policyID              13571 non-null int64
statecode             13571 non-null object
county                13571 non-null object
eq_site_limit         13571 non-null float64
hu_site_limit         13571 non-null float64
fl_site_limit         13571 non-null float64
fr_site_limit         13571 non-null float64
tiv_2011              13571 non-null float64
tiv_2012              13571 non-null float64
eq_site_deductible    13571 non-null float64
hu_site_deductible    13571 non-null float64
fl_site_deductible    13571 non-null float64
fr_site_deductible    13571 non-null int64
point_latitude        13571 non-null float64
point_longitude       13571 non-null float64
line                  13570 non-null object
construction          13570 non-null object
point_granularity     13570 non-null float64
dtypes: float64(12), int64(2), object(4)
memory usage: 1.9+ MB


NoneType

# Accesing a single column

In [17]:
# To access the column policyID
data.policyID

0        119736
1        448094
2        206893
3        333743
4        172534
5        785275
6        995932
7        223488
8        433512
9        142071
10       253816
11       894922
12       422834
13       582721
14       842700
15       874333
16       580146
17       456149
18       767862
19       353022
20       367814
21       671392
22       772887
23       983122
24       934215
25       385951
26       716332
27       751262
28       633663
29       105851
          ...  
13541    604815
13542    123581
13543    830561
13544    567945
13545    304010
13546    202453
13547    920181
13548    186980
13549    350034
13550    292859
13551    286864
13552    820966
13553    756126
13554    617621
13555    950327
13556    796959
13557    270198
13558    809530
13559    474787
13560    260965
13561    613939
13562    508003
13563    638040
13564    826497
13565    137757
13566    906943
13567    758044
13568    387570
13569    310787
13570    522718
Name: policyID, Length: 

In [18]:

# What is the ouput type
print(type(data.policyID))

<class 'pandas.core.series.Series'>


In [19]:
# To access the column policyID
data['policyID']

0        119736
1        448094
2        206893
3        333743
4        172534
5        785275
6        995932
7        223488
8        433512
9        142071
10       253816
11       894922
12       422834
13       582721
14       842700
15       874333
16       580146
17       456149
18       767862
19       353022
20       367814
21       671392
22       772887
23       983122
24       934215
25       385951
26       716332
27       751262
28       633663
29       105851
          ...  
13541    604815
13542    123581
13543    830561
13544    567945
13545    304010
13546    202453
13547    920181
13548    186980
13549    350034
13550    292859
13551    286864
13552    820966
13553    756126
13554    617621
13555    950327
13556    796959
13557    270198
13558    809530
13559    474787
13560    260965
13561    613939
13562    508003
13563    638040
13564    826497
13565    137757
13566    906943
13567    758044
13568    387570
13569    310787
13570    522718
Name: policyID, Length: 

In [20]:
# What is the ouput type
print(type(data['policyID']))

<class 'pandas.core.series.Series'>


# specific column Datatype

In [21]:
data.statecode

0        FL
1        FL
2        FL
3        FL
4        FL
5        FL
6        FL
7        FL
8        FL
9        FL
10       FL
11       FL
12       FL
13       FL
14       FL
15       FL
16       FL
17       FL
18       FL
19       FL
20       FL
21       FL
22       FL
23       FL
24       FL
25       FL
26       FL
27       FL
28       FL
29       FL
         ..
13541    FL
13542    FL
13543    FL
13544    FL
13545    FL
13546    FL
13547    FL
13548    FL
13549    FL
13550    FL
13551    FL
13552    FL
13553    FL
13554    FL
13555    FL
13556    FL
13557    FL
13558    FL
13559    FL
13560    FL
13561    FL
13562    FL
13563    FL
13564    FL
13565    FL
13566    FL
13567    FL
13568    FL
13569    FL
13570    FL
Name: statecode, Length: 13571, dtype: object

In [22]:
# Specific column datatype
data.statecode.dtype

dtype('O')

In [23]:
data.statecode.dtypes

dtype('O')

In [2]:
import pandas as pd
data = pd.read_csv("FL_insurance_sample.csv")
data.dtype

AttributeError: 'DataFrame' object has no attribute 'dtype'

# difference b/n dtype and dtypes

- datatafrme object doesn't have dtype
- Numerical columns - int or float data
- mean,median,mode,sum.quartiles,min,max
- Categorical columns - Object type(string data)
- Bollean data - True or False

In [3]:
data.point_granularity

0        1.0
1        3.0
2        1.0
3        3.0
4        1.0
5        3.0
6        1.0
7        1.0
8        1.0
9        1.0
10       1.0
11       1.0
12       1.0
13       1.0
14       1.0
15       1.0
16       3.0
17       1.0
18       3.0
19       1.0
20       1.0
21       3.0
22       1.0
23       3.0
24       1.0
25       1.0
26       1.0
27       1.0
28       1.0
29       1.0
        ... 
13541    4.0
13542    1.0
13543    1.0
13544    1.0
13545    3.0
13546    1.0
13547    3.0
13548    3.0
13549    3.0
13550    1.0
13551    1.0
13552    1.0
13553    1.0
13554    1.0
13555    1.0
13556    1.0
13557    1.0
13558    1.0
13559    4.0
13560    1.0
13561    1.0
13562    1.0
13563    1.0
13564    3.0
13565    1.0
13566    1.0
13567    1.0
13568    3.0
13569    3.0
13570    NaN
Name: point_granularity, Length: 13571, dtype: float64

In [4]:
data.point_granularity.dtype

dtype('float64')

In [5]:
# mean value of a column
data.point_granularity.mean()

1.7277081798084009

In [6]:
# mean value of a column
data.point_granularity.sum()/data.point_granularity.shape[0]

1.7275808709748728

In [7]:
# minimum value of a column
print(data.point_granularity.min())
# maximum value of a column
print(data.point_granularity.max())

1.0
4.0


In [8]:
data.point_granularity.value_counts()

1.0    9427
3.0    2554
4.0    1589
Name: point_granularity, dtype: int64

In [9]:
data.point_granularity.mode()
# output type is a series

0    1.0
dtype: float64

In [10]:
data.point_granularity.mode()[0]

1.0

In [11]:
data.point_granularity.median()

1.0

In [12]:
# Range of values in a column
(data.point_granularity.min(),data.point_granularity.max())

(1.0, 4.0)

In [13]:
data.point_granularity.count()

13570

In [14]:
data.county.dtype

dtype('O')

In [15]:
# How many non null values are there
data.county.count()
# Output type is an integer

13571

In [16]:
# For each uniwue value what is the count
data.county.value_counts()
# output type is a series
# In that series the labels are unique values
# the value for that label is how many times it occured in that column
# the final dtype refers to the value type of the series. count is integer

DUVAL COUNTY         1894
VOLUSIA COUNTY       1367
OKALOOSA COUNTY      1115
SEMINOLE COUNTY      1100
ORANGE COUNTY        1035
ALACHUA COUNTY        973
SANTA ROSA COUNTY     856
ST  JOHNS COUNTY      657
ESCAMBIA COUNTY       494
BAY COUNTY            403
CLAY COUNTY           363
WALTON COUNTY         288
PUTNAM COUNTY         268
LEON COUNTY           246
JACKSON COUNTY        208
LAKE COUNTY           206
FLAGLER COUNTY        204
GADSDEN COUNTY        196
SUWANNEE COUNTY       154
NASSAU COUNTY         135
COLUMBIA COUNTY       125
WASHINGTON COUNTY     116
LEVY COUNTY           114
TAYLOR COUNTY         113
BREVARD COUNTY         86
WAKULLA COUNTY         85
MADISON COUNTY         81
MARION COUNTY          72
GULF COUNTY            72
BAKER COUNTY           70
LAFAYETTE COUNTY       68
CALHOUN COUNTY         68
JEFFERSON COUNTY       57
HOLMES COUNTY          40
DIXIE COUNTY           40
GILCHRIST COUNTY       39
FRANKLIN COUNTY        37
LIBERTY COUNTY         36
HAMILTON COU

In [17]:

# How many unique values are there
print(data.county.nunique())
print(data.county.value_counts().shape[0])

42
42


In [18]:
# What are the uinique values
print(data.county.unique())
print(data.county.value_counts().index)

['CLAY COUNTY' 'SUWANNEE COUNTY' 'NASSAU COUNTY' 'COLUMBIA COUNTY'
 'ST  JOHNS COUNTY' 'BAKER COUNTY' 'BRADFORD COUNTY' 'HAMILTON COUNTY'
 'UNION COUNTY' 'MADISON COUNTY' 'LAFAYETTE COUNTY' 'FLAGLER COUNTY'
 'DUVAL COUNTY' 'LAKE COUNTY' 'VOLUSIA COUNTY' 'PUTNAM COUNTY'
 'MARION COUNTY' 'SUMTER COUNTY' 'LEON COUNTY' 'FRANKLIN COUNTY'
 'LIBERTY COUNTY' 'GADSDEN COUNTY' 'WAKULLA COUNTY' 'JEFFERSON COUNTY'
 'TAYLOR COUNTY' 'BAY COUNTY' 'WALTON COUNTY' 'JACKSON COUNTY'
 'CALHOUN COUNTY' 'HOLMES COUNTY' 'WASHINGTON COUNTY' 'GULF COUNTY'
 'ESCAMBIA COUNTY' 'SANTA ROSA COUNTY' 'OKALOOSA COUNTY' 'ALACHUA COUNTY'
 'GILCHRIST COUNTY' 'LEVY COUNTY' 'DIXIE COUNTY' 'SEMINOLE COUNTY'
 'ORANGE COUNTY' 'BREVARD COUNTY']
Index(['DUVAL COUNTY', 'VOLUSIA COUNTY', 'OKALOOSA COUNTY', 'SEMINOLE COUNTY',
       'ORANGE COUNTY', 'ALACHUA COUNTY', 'SANTA ROSA COUNTY',
       'ST  JOHNS COUNTY', 'ESCAMBIA COUNTY', 'BAY COUNTY', 'CLAY COUNTY',
       'WALTON COUNTY', 'PUTNAM COUNTY', 'LEON COUNTY', 'JACKSON COUNT

In [19]:
# To get the description for the numerical columns
data.describe()

policyID  eq_site_limit  hu_site_limit  fl_site_limit  \
count   13571.000000   1.357100e+04   1.357100e+04   1.357100e+04   
mean   549845.485226   5.942668e+05   1.444358e+06   5.724015e+05   
std    261073.957434   1.196783e+07   1.337212e+07   1.192754e+07   
min    100141.000000   0.000000e+00   0.000000e+00   0.000000e+00   
25%    322686.500000   0.000000e+00   2.022421e+04   0.000000e+00   
50%    548825.000000   0.000000e+00   1.254999e+05   0.000000e+00   
75%    776374.000000   5.647500e+03   6.345000e+05   3.478500e+03   
max    999971.000000   1.017000e+09   1.017000e+09   1.017000e+09   

       fr_site_limit      tiv_2011      tiv_2012  eq_site_deductible  \
count   1.357100e+04  1.357100e+04  1.357100e+04        1.357100e+04   
mean    7.876392e+05  1.545899e+06  1.802276e+06        1.826912e+03   
std     1.204500e+07  1.365932e+07  1.371633e+07        7.821441e+04   
min     0.000000e+00  9.000000e+01  7.601000e+01        0.000000e+00   
25%     0.000000e+00  2.439088e+04  2.826878e+04        0.000000e+00   
50%     0.000000e+00  1.346717e+05  1.567735e+05        0.000000e+00   
75%     4.410900e+04  6.682196e+05  7.899712e+05        0.000000e+00   
max     1.017000e+09  1.017000e+09  8.883000e+08        6.273765e+06   

       hu_site_deductible  fl_site_deductible  fr_site_deductible  \
count        1.357100e+04        13571.000000        13571.000000   
mean         5.792175e+03          184.195866            0.895291   
std          8.847850e+04         7914.001120           32.075838   
min          0.000000e+00            0.000000            0.000000   
25%          0.000000e+00            0.000000            0.000000   
50%          0.000000e+00            0.000000            0.000000   
75%          0.000000e+00            0.000000            0.000000   
max          6.273765e+06       450000.000000         2250.000000   

       point_latitude  point_longitude  point_granularity  
count    13571.000000     13571.000000       13570.000000  
mean        29.838036       -83.097801           1.727708  
std          0.728318         2.212719           1.130151  
min         28.347080       -87.447290           1.000000  
25%         29.149385       -85.291995           1.000000  
50%         30.140441       -81.738100           1.000000  
75%         30.411170       -81.381046           3.000000  
max         30.989820       -80.790880           4.000000

In [20]:
# To get the list of numerical columns
data.describe().columns

Index(['policyID', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit',
       'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible',
       'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible',
       'point_latitude', 'point_longitude', 'point_granularity'],
      dtype='object')

In [21]:
# To get the description for the categorical columns
data.describe(include=object)

statecode        county         line construction
count      13571         13571        13570        13570
unique         1            42            2            5
top           FL  DUVAL COUNTY  Residential         Wood
freq       13571          1894        11944         8940

In [23]:
import numpy as np
data.describe(include=np.int64)

policyID  fr_site_deductible
count   13571.000000        13571.000000
mean   549845.485226            0.895291
std    261073.957434           32.075838
min    100141.000000            0.000000
25%    322686.500000            0.000000
50%    548825.000000            0.000000
75%    776374.000000            0.000000
max    999971.000000         2250.000000

In [24]:
data.describe(include=np.float64)

eq_site_limit  hu_site_limit  fl_site_limit  fr_site_limit  \
count   1.357100e+04   1.357100e+04   1.357100e+04   1.357100e+04   
mean    5.942668e+05   1.444358e+06   5.724015e+05   7.876392e+05   
std     1.196783e+07   1.337212e+07   1.192754e+07   1.204500e+07   
min     0.000000e+00   0.000000e+00   0.000000e+00   0.000000e+00   
25%     0.000000e+00   2.022421e+04   0.000000e+00   0.000000e+00   
50%     0.000000e+00   1.254999e+05   0.000000e+00   0.000000e+00   
75%     5.647500e+03   6.345000e+05   3.478500e+03   4.410900e+04   
max     1.017000e+09   1.017000e+09   1.017000e+09   1.017000e+09   

           tiv_2011      tiv_2012  eq_site_deductible  hu_site_deductible  \
count  1.357100e+04  1.357100e+04        1.357100e+04        1.357100e+04   
mean   1.545899e+06  1.802276e+06        1.826912e+03        5.792175e+03   
std    1.365932e+07  1.371633e+07        7.821441e+04        8.847850e+04   
min    9.000000e+01  7.601000e+01        0.000000e+00        0.000000e+00   
25%    2.439088e+04  2.826878e+04        0.000000e+00        0.000000e+00   
50%    1.346717e+05  1.567735e+05        0.000000e+00        0.000000e+00   
75%    6.682196e+05  7.899712e+05        0.000000e+00        0.000000e+00   
max    1.017000e+09  8.883000e+08        6.273765e+06        6.273765e+06   

       fl_site_deductible  point_latitude  point_longitude  point_granularity  
count        13571.000000    13571.000000     13571.000000       13570.000000  
mean           184.195866       29.838036       -83.097801           1.727708  
std           7914.001120        0.728318         2.212719           1.130151  
min              0.000000       28.347080       -87.447290           1.000000  
25%              0.000000       29.149385       -85.291995           1.000000  
50%              0.000000       30.140441       -81.738100           1.000000  
75%              0.000000       30.411170       -81.381046           3.000000  
max         450000.000000       30.989820       -80.790880           4.000000

In [25]:
data.describe(include='all')

policyID statecode        county  eq_site_limit  hu_site_limit  \
count    13571.000000     13571         13571   1.357100e+04   1.357100e+04   
unique            NaN         1            42            NaN            NaN   
top               NaN        FL  DUVAL COUNTY            NaN            NaN   
freq              NaN     13571          1894            NaN            NaN   
mean    549845.485226       NaN           NaN   5.942668e+05   1.444358e+06   
std     261073.957434       NaN           NaN   1.196783e+07   1.337212e+07   
min     100141.000000       NaN           NaN   0.000000e+00   0.000000e+00   
25%     322686.500000       NaN           NaN   0.000000e+00   2.022421e+04   
50%     548825.000000       NaN           NaN   0.000000e+00   1.254999e+05   
75%     776374.000000       NaN           NaN   5.647500e+03   6.345000e+05   
max     999971.000000       NaN           NaN   1.017000e+09   1.017000e+09   

        fl_site_limit  fr_site_limit      tiv_2011      tiv_2012  \
count    1.357100e+04   1.357100e+04  1.357100e+04  1.357100e+04   
unique            NaN            NaN           NaN           NaN   
top               NaN            NaN           NaN           NaN   
freq              NaN            NaN           NaN           NaN   
mean     5.724015e+05   7.876392e+05  1.545899e+06  1.802276e+06   
std      1.192754e+07   1.204500e+07  1.365932e+07  1.371633e+07   
min      0.000000e+00   0.000000e+00  9.000000e+01  7.601000e+01   
25%      0.000000e+00   0.000000e+00  2.439088e+04  2.826878e+04   
50%      0.000000e+00   0.000000e+00  1.346717e+05  1.567735e+05   
75%      3.478500e+03   4.410900e+04  6.682196e+05  7.899712e+05   
max      1.017000e+09   1.017000e+09  1.017000e+09  8.883000e+08   

        eq_site_deductible  hu_site_deductible  fl_site_deductible  \
count         1.357100e+04        1.357100e+04        13571.000000   
unique                 NaN                 NaN                 NaN   
top                    NaN                 NaN                 NaN   
freq                   NaN                 NaN                 NaN   
mean          1.826912e+03        5.792175e+03          184.195866   
std           7.821441e+04        8.847850e+04         7914.001120   
min           0.000000e+00        0.000000e+00            0.000000   
25%           0.000000e+00        0.000000e+00            0.000000   
50%           0.000000e+00        0.000000e+00            0.000000   
75%           0.000000e+00        0.000000e+00            0.000000   
max           6.273765e+06        6.273765e+06       450000.000000   

        fr_site_deductible  point_latitude  point_longitude         line  \
count         13571.000000    13571.000000     13571.000000        13570   
unique                 NaN             NaN              NaN            2   
top                    NaN             NaN              NaN  Residential   
freq                   NaN             NaN              NaN        11944   
mean              0.895291       29.838036       -83.097801          NaN   
std              32.075838        0.728318         2.212719          NaN   
min               0.000000       28.347080       -87.447290          NaN   
25%               0.000000       29.149385       -85.291995          NaN   
50%               0.000000       30.140441       -81.738100          NaN   
75%               0.000000       30.411170       -81.381046          NaN   
max            2250.000000       30.989820       -80.790880          NaN   

       construction  point_granularity  
count         13570       13570.000000  
unique            5                NaN  
top            Wood                NaN  
freq           8940                NaN  
mean            NaN           1.727708  
std             NaN           1.130151  
min             NaN           1.000000  
25%             NaN           1.000000  
50%             NaN           1.000000  
75%             NaN           3.000000  
max             NaN           4.000000

In [26]:
data.columns

Index(['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit',
       'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012',
       'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible',
       'fr_site_deductible', 'point_latitude', 'point_longitude', 'line',
       'construction', 'point_granularity'],
      dtype='object')

# Slicing, subsetting, Boolean Indexing

- postion based (iloc)
- rows --> 0 to 36633
- columns --> 0 to 17
- label based indexing(loc)
- rows --> 0 to 36633
- columns --> ['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible', 'point_latitude', 'point_longitude', 'line', 'construction', 'point_granularity']

In [27]:
data.iloc[1,:] # row with index 1 and all the columns

policyID                   448094
statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 1, dtype: object

In [28]:
data.iloc[1,1] # row with index 1and column with index 1

'FL'

In [29]:
data.loc[1,"statecode"]

'FL'

In [30]:
data.iloc[1,1:3] # row with index 1 and columns with index 1,2
# iloc stop index is exclusive

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [31]:

data.loc[1,"statecode":"county"] 
# loc - the stop label also inclusive

statecode             FL
county       CLAY COUNTY
Name: 1, dtype: object

In [32]:
data.iloc[1:4,1:3] 
# rows with indices 1,2,3
# columns with indices 1,2

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY

In [33]:
data.loc[1:3,"statecode":"county"]

statecode       county
1        FL  CLAY COUNTY
2        FL  CLAY COUNTY
3        FL  CLAY COUNTY

In [34]:
data.policyID.nunique()

13571

In [35]:
data.set_index("policyID")

statecode         county  eq_site_limit  hu_site_limit  \
policyID                                                          
119736          FL    CLAY COUNTY       498960.0      498960.00   
448094          FL    CLAY COUNTY      1322376.3     1322376.30   
206893          FL    CLAY COUNTY       190724.4      190724.40   
333743          FL    CLAY COUNTY            0.0       79520.76   
172534          FL    CLAY COUNTY            0.0      254281.50   
785275          FL    CLAY COUNTY            0.0      515035.62   
995932          FL    CLAY COUNTY            0.0    19260000.00   
223488          FL    CLAY COUNTY       328500.0      328500.00   
433512          FL    CLAY COUNTY       315000.0      315000.00   
142071          FL    CLAY COUNTY       705600.0      705600.00   
253816          FL    CLAY COUNTY       831498.3      831498.30   
894922          FL    CLAY COUNTY            0.0       24059.09   
422834          FL    CLAY COUNTY            0.0       48115.94   
582721          FL    CLAY COUNTY            0.0       28869.12   
842700          FL    CLAY COUNTY            0.0       56135.64   
874333          FL    CLAY COUNTY            0.0       48115.94   
580146          FL    CLAY COUNTY            0.0       48115.94   
456149          FL    CLAY COUNTY            0.0       80192.49   
767862          FL    CLAY COUNTY            0.0       48115.94   
353022          FL    CLAY COUNTY            0.0       60946.79   
367814          FL    CLAY COUNTY            0.0       28869.12   
671392          FL    CLAY COUNTY            0.0    13410000.00   
772887          FL    CLAY COUNTY            0.0     1669113.93   
983122          FL    CLAY COUNTY            0.0      179562.23   
934215          FL    CLAY COUNTY            0.0      177744.16   
385951          FL    CLAY COUNTY            0.0       17757.58   
716332          FL    CLAY COUNTY            0.0      130129.87   
751262          FL    CLAY COUNTY            0.0       42854.77   
633663          FL    CLAY COUNTY            0.0         785.58   
105851          FL    CLAY COUNTY            0.0      170361.91   
...            ...            ...            ...            ...   
604815          FL  ORANGE COUNTY       497718.0      497718.00   
123581          FL  ORANGE COUNTY      1083600.0     1083600.00   
830561          FL  ORANGE COUNTY       418230.0      418230.00   
567945          FL  ORANGE COUNTY            0.0      277276.92   
304010          FL  ORANGE COUNTY            0.0    14130000.00   
202453          FL  ORANGE COUNTY            0.0    19260000.00   
920181          FL  ORANGE COUNTY            0.0    19620000.00   
186980          FL  ORANGE COUNTY            0.0     8432564.40   
350034          FL  ORANGE COUNTY            0.0     2858963.40   
292859          FL  ORANGE COUNTY            0.0    19620000.00   
286864          FL  ORANGE COUNTY            0.0      681474.42   
820966          FL  ORANGE COUNTY            0.0     3941825.58   
756126          FL  ORANGE COUNTY            0.0      190479.96   
617621          FL  ORANGE COUNTY            0.0      563467.50   
950327          FL  ORANGE COUNTY            0.0     1905773.72   
796959          FL  ORANGE COUNTY            0.0     1391883.85   
270198          FL  ORANGE COUNTY            0.0     4534866.90   
809530          FL  ORANGE COUNTY            0.0     1656192.60   
474787          FL  ORANGE COUNTY     10530000.0    10530000.00   
260965          FL  ORANGE COUNTY        27000.0       27000.00   
613939          FL  ORANGE COUNTY            0.0     8094677.30   
508003          FL  ORANGE COUNTY            0.0     8050879.41   
638040          FL  ORANGE COUNTY       886592.7      886592.70   
826497          FL  ORANGE COUNTY       226649.7      226649.70   
137757          FL  ORANGE COUNTY       923010.3      923010.30   
906943          FL  ORANGE COUNTY            0.0     1566401.56   
758044          FL  ORANGE COUNTY      4791330.0     479

In [36]:
data

policyID statecode         county  eq_site_limit  hu_site_limit  \
0        119736        FL    CLAY COUNTY       498960.0      498960.00   
1        448094        FL    CLAY COUNTY      1322376.3     1322376.30   
2        206893        FL    CLAY COUNTY       190724.4      190724.40   
3        333743        FL    CLAY COUNTY            0.0       79520.76   
4        172534        FL    CLAY COUNTY            0.0      254281.50   
5        785275        FL    CLAY COUNTY            0.0      515035.62   
6        995932        FL    CLAY COUNTY            0.0    19260000.00   
7        223488        FL    CLAY COUNTY       328500.0      328500.00   
8        433512        FL    CLAY COUNTY       315000.0      315000.00   
9        142071        FL    CLAY COUNTY       705600.0      705600.00   
10       253816        FL    CLAY COUNTY       831498.3      831498.30   
11       894922        FL    CLAY COUNTY            0.0       24059.09   
12       422834        FL    CLAY COUNTY            0.0       48115.94   
13       582721        FL    CLAY COUNTY            0.0       28869.12   
14       842700        FL    CLAY COUNTY            0.0       56135.64   
15       874333        FL    CLAY COUNTY            0.0       48115.94   
16       580146        FL    CLAY COUNTY            0.0       48115.94   
17       456149        FL    CLAY COUNTY            0.0       80192.49   
18       767862        FL    CLAY COUNTY            0.0       48115.94   
19       353022        FL    CLAY COUNTY            0.0       60946.79   
20       367814        FL    CLAY COUNTY            0.0       28869.12   
21       671392        FL    CLAY COUNTY            0.0    13410000.00   
22       772887        FL    CLAY COUNTY            0.0     1669113.93   
23       983122        FL    CLAY COUNTY            0.0      179562.23   
24       934215        FL    CLAY COUNTY            0.0      177744.16   
25       385951        FL    CLAY COUNTY            0.0       17757.58   
26       716332        FL    CLAY COUNTY            0.0      130129.87   
27       751262        FL    CLAY COUNTY            0.0       42854.77   
28       633663        FL    CLAY COUNTY            0.0         785.58   
29       105851        FL    CLAY COUNTY            0.0      170361.91   
...         ...       ...            ...            ...            ...   
13541    604815        FL  ORANGE COUNTY       497718.0      497718.00   
13542    123581        FL  ORANGE COUNTY      1083600.0     1083600.00   
13543    830561        FL  ORANGE COUNTY       418230.0      418230.00   
13544    567945        FL  ORANGE COUNTY            0.0      277276.92   
13545    304010        FL  ORANGE COUNTY            0.0    14130000.00   
13546    202453        FL  ORANGE COUNTY            0.0    19260000.00   
13547    920181        FL  ORANGE COUNTY            0.0    19620000.00   
13548    186980        FL  ORANGE COUNTY            0.0     8432564.40   
13549    350034        FL  ORANGE COUNTY            0.0     2858963.40   
13550    292859        FL  ORANGE COUNTY            0.0    19620000.00   
13551    286864        FL  ORANGE COUNTY            0.0      681474.42   
13552    820966        FL  ORANGE COUNTY            0.0     3941825.58   
13553    756126        FL  ORANGE COUNTY            0.0      190479.96   
13554    617621        FL  ORANGE COUNTY            0.0      563467.50   
13555    950327        FL  ORANGE COUNTY            0.0     1905773.72   
13556    796959        FL  ORANGE COUNTY            0.0     1391883.85   
13557    270198        FL  ORANGE COUNTY            0.0     4534866.90   
13558    809530        FL  ORANGE COUNTY            0.0     1656192.60   
13559    474787        FL  ORANGE COUNTY     10530000.0    10530000.00   
13560    260965        FL  ORANGE COUNTY        27000.0       27000.00   
13561    613939        FL  ORANGE COUNTY            0.0     8094677.30   
13562    508003        FL  ORANGE COUNTY            0.0     8050879.41   
13563    63

In [37]:
df = data.set_index("policyID")
df.head()

statecode       county  eq_site_limit  hu_site_limit  fl_site_limit  \
policyID                                                                       
119736          FL  CLAY COUNTY       498960.0      498960.00       498960.0   
448094          FL  CLAY COUNTY      1322376.3     1322376.30      1322376.3   
206893          FL  CLAY COUNTY       190724.4      190724.40       190724.4   
333743          FL  CLAY COUNTY            0.0       79520.76            0.0   
172534          FL  CLAY COUNTY            0.0      254281.50            0.0   

          fr_site_limit    tiv_2011    tiv_2012  eq_site_deductible  \
policyID                                                              
119736         498960.0   498960.00   792148.90                 0.0   
448094        1322376.3  1322376.30  1438163.57                 0.0   
206893         190724.4   190724.40   192476.78                 0.0   
333743              0.0    79520.76    86854.48                 0.0   
172534         254281.5   254281.50   246144.49                 0.0   

          hu_site_deductible  fl_site_deductible  fr_site_deductible  \
policyID                                                               
119736                9979.2                 0.0                   0   
448094                   0.0                 0.0                   0   
206893                   0.0                 0.0                   0   
333743                   0.0                 0.0                   0   
172534                   0.0                 0.0                   0   

          point_latitude  point_longitude         line construction  \
policyID                                                              
119736         30.102261       -81.711777  Residential      Masonry   
448094         30.063936       -81.707664  Residential      Masonry   
206893         30.089579       -81.700455  Residential         Wood   
333743         30.063236       -81.707703  Residential         Wood   
172534         30.060614       -81.702675  Residential         Wood   

          point_granularity  
policyID                     
119736                  1.0  
448094                  3.0  
206893                  1.0  
333743                  3.0  
172534                  1.0

In [38]:
df.index

Int64Index([119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488,
            433512, 142071,
            ...
            613939, 508003, 638040, 826497, 137757, 906943, 758044, 387570,
            310787, 522718],
           dtype='int64', name='policyID', length=13571)

# Slicing, subsetting, Boolean Indexing with df

- postion based (iloc)
- rows --> 0 to 36633
- columns --> 0 to 17
- label based indexing(loc)
- rows --> [119736, 448094, 206893, 333743, 172534, 785275, 995932, 223488,
  433512, 142071,...]
- columns --> ['policyID', 'statecode', 'county', 'eq_site_limit', 'hu_site_limit', 'fl_site_limit', 'fr_site_limit', 'tiv_2011', 'tiv_2012', 'eq_site_deductible', 'hu_site_deductible', 'fl_site_deductible', 'fr_site_deductible', 'point_latitude', 'point_longitude', 'line', 'construction', 'point_granularity']


In [40]:
# row with index 1 and all the columns
df.iloc[1,:]

statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 448094, dtype: object

In [41]:
df.loc[448094,:]

statecode                      FL
county                CLAY COUNTY
eq_site_limit         1.32238e+06
hu_site_limit         1.32238e+06
fl_site_limit         1.32238e+06
fr_site_limit         1.32238e+06
tiv_2011              1.32238e+06
tiv_2012              1.43816e+06
eq_site_deductible              0
hu_site_deductible              0
fl_site_deductible              0
fr_site_deductible              0
point_latitude            30.0639
point_longitude          -81.7077
line                  Residential
construction              Masonry
point_granularity               3
Name: 448094, dtype: object

In [42]:
# row with index 1and column with index 1
df.iloc[1,1]

'CLAY COUNTY'

In [43]:
df.loc[448094,"county"]

'CLAY COUNTY'

In [44]:
# row with index 1 and columns with index 1,2
df.iloc[1,1:3]


county           CLAY COUNTY
eq_site_limit    1.32238e+06
Name: 448094, dtype: object

In [45]:
df.loc[ 448094,"county":"eq_site_limit"]

county           CLAY COUNTY
eq_site_limit    1.32238e+06
Name: 448094, dtype: object

In [46]:
# rows with indices 1,2,3
# columns with indices 1,2
df.iloc[1:4,:]

statecode       county  eq_site_limit  hu_site_limit  fl_site_limit  \
policyID                                                                       
448094          FL  CLAY COUNTY      1322376.3     1322376.30      1322376.3   
206893          FL  CLAY COUNTY       190724.4      190724.40       190724.4   
333743          FL  CLAY COUNTY            0.0       79520.76            0.0   

          fr_site_limit    tiv_2011    tiv_2012  eq_site_deductible  \
policyID                                                              
448094        1322376.3  1322376.30  1438163.57                 0.0   
206893         190724.4   190724.40   192476.78                 0.0   
333743              0.0    79520.76    86854.48                 0.0   

          hu_site_deductible  fl_site_deductible  fr_site_deductible  \
policyID                                                               
448094                   0.0                 0.0                   0   
206893                   0.0                 0.0                   0   
333743                   0.0                 0.0                   0   

          point_latitude  point_longitude         line construction  \
policyID                                                              
448094         30.063936       -81.707664  Residential      Masonry   
206893         30.089579       -81.700455  Residential         Wood   
333743         30.063236       -81.707703  Residential         Wood   

          point_granularity  
policyID                     
448094                  3.0  
206893                  1.0  
333743                  3.0

In [47]:
df.loc[448094:333743,:]

statecode       county  eq_site_limit  hu_site_limit  fl_site_limit  \
policyID                                                                       
448094          FL  CLAY COUNTY      1322376.3     1322376.30      1322376.3   
206893          FL  CLAY COUNTY       190724.4      190724.40       190724.4   
333743          FL  CLAY COUNTY            0.0       79520.76            0.0   

          fr_site_limit    tiv_2011    tiv_2012  eq_site_deductible  \
policyID                                                              
448094        1322376.3  1322376.30  1438163.57                 0.0   
206893         190724.4   190724.40   192476.78                 0.0   
333743              0.0    79520.76    86854.48                 0.0   

          hu_site_deductible  fl_site_deductible  fr_site_deductible  \
policyID                                                               
448094                   0.0                 0.0                   0   
206893                   0.0                 0.0                   0   
333743                   0.0                 0.0                   0   

          point_latitude  point_longitude         line construction  \
policyID                                                              
448094         30.063936       -81.707664  Residential      Masonry   
206893         30.089579       -81.700455  Residential         Wood   
333743         30.063236       -81.707703  Residential         Wood   

          point_granularity  
policyID                     
448094                  3.0  
206893                  1.0  
333743                  3.0

# Boolean Indexing

In [48]:
df.dtypes

statecode              object
county                 object
eq_site_limit         float64
hu_site_limit         float64
fl_site_limit         float64
fr_site_limit         float64
tiv_2011              float64
tiv_2012              float64
eq_site_deductible    float64
hu_site_deductible    float64
fl_site_deductible    float64
fr_site_deductible      int64
point_latitude        float64
point_longitude       float64
line                   object
construction           object
point_granularity     float64
dtype: object

In [49]:
df.dtypes == object

statecode              True
county                 True
eq_site_limit         False
hu_site_limit         False
fl_site_limit         False
fr_site_limit         False
tiv_2011              False
tiv_2012              False
eq_site_deductible    False
hu_site_deductible    False
fl_site_deductible    False
fr_site_deductible    False
point_latitude        False
point_longitude       False
line                   True
construction           True
point_granularity     False
dtype: bool

In [50]:
df.loc[:,df.dtypes == object]

statecode         county         line         construction
policyID                                                           
119736          FL    CLAY COUNTY  Residential              Masonry
448094          FL    CLAY COUNTY  Residential              Masonry
206893          FL    CLAY COUNTY  Residential                 Wood
333743          FL    CLAY COUNTY  Residential                 Wood
172534          FL    CLAY COUNTY  Residential                 Wood
785275          FL    CLAY COUNTY  Residential              Masonry
995932          FL    CLAY COUNTY   Commercial  Reinforced Concrete
223488          FL    CLAY COUNTY  Residential                 Wood
433512          FL    CLAY COUNTY  Residential                 Wood
142071          FL    CLAY COUNTY  Residential              Masonry
253816          FL    CLAY COUNTY  Residential              Masonry
894922          FL    CLAY COUNTY  Residential                 Wood
422834          FL    CLAY COUNTY  Residential                 Wood
582721          FL    CLAY COUNTY  Residential                 Wood
842700          FL    CLAY COUNTY  Residential                 Wood
874333          FL    CLAY COUNTY  Residential                 Wood
580146          FL    CLAY COUNTY  Residential                 Wood
456149          FL    CLAY COUNTY  Residential                 Wood
767862          FL    CLAY COUNTY  Residential                 Wood
353022          FL    CLAY COUNTY  Residential                 Wood
367814          FL    CLAY COUNTY  Residential                 Wood
671392          FL    CLAY COUNTY   Commercial  Reinforced Concrete
772887          FL    CLAY COUNTY  Residential              Masonry
983122          FL    CLAY COUNTY  Residential                 Wood
934215          FL    CLAY COUNTY  Residential                 Wood
385951          FL    CLAY COUNTY  Residential                 Wood
716332          FL    CLAY COUNTY  Residential                 Wood
751262          FL    CLAY COUNTY  Residential                 Wood
633663          FL    CLAY COUNTY  Residential                 Wood
105851          FL    CLAY COUNTY  Residential                 Wood
...            ...            ...          ...                  ...
604815          FL  ORANGE COUNTY  Residential              Masonry
123581          FL  ORANGE COUNTY  Residential              Masonry
830561          FL  ORANGE COUNTY  Residential              Masonry
567945          FL  ORANGE COUNTY  Residential                 Wood
304010          FL  ORANGE COUNTY   Commercial  Reinforced Concrete
202453          FL  ORANGE COUNTY   Commercial  Reinforced Concrete
920181          FL  ORANGE COUNTY   Commercial  Reinforced Concrete
186980          FL  ORANGE COUNTY   Commercial   Reinforced Masonry
350034          FL  ORANGE COUNTY   Commercial   Reinforced Masonry
292859          FL  ORANGE COUNTY   Commercial  Reinforced Concrete
286864          FL  ORANGE COUNTY  Residential              Masonry
820966          FL  ORANGE COUNTY   Commercial   Reinforced Masonry
756126          FL  ORANGE COUNTY  Residential                 Wood
617621          FL  ORANGE COUNTY  Residential              Masonry
950327          FL  ORANGE COUNTY   Commercial   Reinforced Masonry
796959          FL  ORANGE COUNTY  Residential              Masonry
270198          FL  ORANGE COUNTY   Commercial   Reinforced Masonry
809530          FL  ORANGE COUNTY  Residential              Masonry
474787          FL  ORANGE COUNTY   Commercial  Reinforced Concrete
260965          FL  ORANGE COUNTY  Residential                 Wood
613939          FL  ORANGE COUNTY   Commercial   Reinforced Masonry
508003          FL  ORANGE COUNTY   Commercial   Reinforced Masonry
638040          FL  ORANGE COUNTY  Residential              Masonry
826497          FL  ORANGE COUNTY  Residential                 Wood
137757          FL  ORANGE COUNTY  Residential              Masonry
906943          FL  ORANGE COUNTY  Residential              Mason

In [51]:
df.dtypes ==np.int64

statecode             False
county                False
eq_site_limit         False
hu_site_limit         False
fl_site_limit         False
fr_site_limit         False
tiv_2011              False
tiv_2012              False
eq_site_deductible    False
hu_site_deductible    False
fl_site_deductible    False
fr_site_deductible     True
point_latitude        False
point_longitude       False
line                  False
construction          False
point_granularity     False
dtype: bool

In [52]:
df.loc[:,df.dtypes == np.int64]

fr_site_deductible
policyID                    
119736                     0
448094                     0
206893                     0
333743                     0
172534                     0
785275                     0
995932                     0
223488                     0
433512                     0
142071                     0
253816                     0
894922                     0
422834                     0
582721                     0
842700                     0
874333                     0
580146                     0
456149                     0
767862                     0
353022                     0
367814                     0
671392                     0
772887                     0
983122                     0
934215                     0
385951                     0
716332                     0
751262                     0
633663                     0
105851                     0
...                      ...
604815                     0
123581                     0
830561                     0
567945                     0
304010                     0
202453                     0
920181                     0
186980                     0
350034                     0
292859                     0
286864                     0
820966                     0
756126                     0
617621                     0
950327                     0
796959                     0
270198                     0
809530                     0
474787                     0
260965                     0
613939                     0
508003                     0
638040                     0
826497                     0
137757                     0
906943                     0
758044                     0
387570                     0
310787                     0
522718                     0

[13571 rows x 1 columns]

In [53]:
df.county

policyID
119736      CLAY COUNTY
448094      CLAY COUNTY
206893      CLAY COUNTY
333743      CLAY COUNTY
172534      CLAY COUNTY
785275      CLAY COUNTY
995932      CLAY COUNTY
223488      CLAY COUNTY
433512      CLAY COUNTY
142071      CLAY COUNTY
253816      CLAY COUNTY
894922      CLAY COUNTY
422834      CLAY COUNTY
582721      CLAY COUNTY
842700      CLAY COUNTY
874333      CLAY COUNTY
580146      CLAY COUNTY
456149      CLAY COUNTY
767862      CLAY COUNTY
353022      CLAY COUNTY
367814      CLAY COUNTY
671392      CLAY COUNTY
772887      CLAY COUNTY
983122      CLAY COUNTY
934215      CLAY COUNTY
385951      CLAY COUNTY
716332      CLAY COUNTY
751262      CLAY COUNTY
633663      CLAY COUNTY
105851      CLAY COUNTY
              ...      
604815    ORANGE COUNTY
123581    ORANGE COUNTY
830561    ORANGE COUNTY
567945    ORANGE COUNTY
304010    ORANGE COUNTY
202453    ORANGE COUNTY
920181    ORANGE COUNTY
186980    ORANGE COUNTY
350034    ORANGE COUNTY
292859    ORANGE COUNTY
286864 

In [54]:
df.county.nunique()

42

In [55]:
df.county.unique()

array(['CLAY COUNTY', 'SUWANNEE COUNTY', 'NASSAU COUNTY',
       'COLUMBIA COUNTY', 'ST  JOHNS COUNTY', 'BAKER COUNTY',
       'BRADFORD COUNTY', 'HAMILTON COUNTY', 'UNION COUNTY',
       'MADISON COUNTY', 'LAFAYETTE COUNTY', 'FLAGLER COUNTY',
       'DUVAL COUNTY', 'LAKE COUNTY', 'VOLUSIA COUNTY', 'PUTNAM COUNTY',
       'MARION COUNTY', 'SUMTER COUNTY', 'LEON COUNTY', 'FRANKLIN COUNTY',
       'LIBERTY COUNTY', 'GADSDEN COUNTY', 'WAKULLA COUNTY',
       'JEFFERSON COUNTY', 'TAYLOR COUNTY', 'BAY COUNTY', 'WALTON COUNTY',
       'JACKSON COUNTY', 'CALHOUN COUNTY', 'HOLMES COUNTY',
       'WASHINGTON COUNTY', 'GULF COUNTY', 'ESCAMBIA COUNTY',
       'SANTA ROSA COUNTY', 'OKALOOSA COUNTY', 'ALACHUA COUNTY',
       'GILCHRIST COUNTY', 'LEVY COUNTY', 'DIXIE COUNTY',
       'SEMINOLE COUNTY', 'ORANGE COUNTY', 'BREVARD COUNTY'], dtype=object)

In [56]:
df.county== 'CLAY COUNTY'

policyID
119736     True
448094     True
206893     True
333743     True
172534     True
785275     True
995932     True
223488     True
433512     True
142071     True
253816     True
894922     True
422834     True
582721     True
842700     True
874333     True
580146     True
456149     True
767862     True
353022     True
367814     True
671392     True
772887     True
983122     True
934215     True
385951     True
716332     True
751262     True
633663     True
105851     True
          ...  
604815    False
123581    False
830561    False
567945    False
304010    False
202453    False
920181    False
186980    False
350034    False
292859    False
286864    False
820966    False
756126    False
617621    False
950327    False
796959    False
270198    False
809530    False
474787    False
260965    False
613939    False
508003    False
638040    False
826497    False
137757    False
906943    False
758044    False
387570    False
310787    False
522718    False
Name: county, L

In [57]:
df.loc[df.county== 'CLAY COUNTY',:]

statecode       county  eq_site_limit  hu_site_limit  fl_site_limit  \
policyID                                                                       
119736          FL  CLAY COUNTY       498960.0      498960.00       498960.0   
448094          FL  CLAY COUNTY      1322376.3     1322376.30      1322376.3   
206893          FL  CLAY COUNTY       190724.4      190724.40       190724.4   
333743          FL  CLAY COUNTY            0.0       79520.76            0.0   
172534          FL  CLAY COUNTY            0.0      254281.50            0.0   
785275          FL  CLAY COUNTY            0.0      515035.62            0.0   
995932          FL  CLAY COUNTY            0.0    19260000.00            0.0   
223488          FL  CLAY COUNTY       328500.0      328500.00       328500.0   
433512          FL  CLAY COUNTY       315000.0      315000.00       315000.0   
142071          FL  CLAY COUNTY       705600.0      705600.00       705600.0   
253816          FL  CLAY COUNTY       831498.3      831498.30       831498.3   
894922          FL  CLAY COUNTY            0.0       24059.09            0.0   
422834          FL  CLAY COUNTY            0.0       48115.94            0.0   
582721          FL  CLAY COUNTY            0.0       28869.12            0.0   
842700          FL  CLAY COUNTY            0.0       56135.64            0.0   
874333          FL  CLAY COUNTY            0.0       48115.94            0.0   
580146          FL  CLAY COUNTY            0.0       48115.94            0.0   
456149          FL  CLAY COUNTY            0.0       80192.49            0.0   
767862          FL  CLAY COUNTY            0.0       48115.94            0.0   
353022          FL  CLAY COUNTY            0.0       60946.79            0.0   
367814          FL  CLAY COUNTY            0.0       28869.12            0.0   
671392          FL  CLAY COUNTY            0.0    13410000.00            0.0   
772887          FL  CLAY COUNTY            0.0     1669113.93            0.0   
983122          FL  CLAY COUNTY            0.0      179562.23            0.0   
934215          FL  CLAY COUNTY            0.0      177744.16            0.0   
385951          FL  CLAY COUNTY            0.0       17757.58            0.0   
716332          FL  CLAY COUNTY            0.0      130129.87            0.0   
751262          FL  CLAY COUNTY            0.0       42854.77            0.0   
633663          FL  CLAY COUNTY            0.0         785.58            0.0   
105851          FL  CLAY COUNTY            0.0      170361.91            0.0   
...            ...          ...            ...            ...            ...   
829730          FL  CLAY COUNTY            0.0    14400000.00            0.0   
693967          FL  CLAY COUNTY            0.0      904575.01            0.0   
780380          FL  CLAY COUNTY            0.0     2485482.37            0.0   
816959          FL  CLAY COUNTY            0.0       47135.08            0.0   
575231          FL  CLAY COUNTY            0.0      436526.99            0.0   
625865          FL  CLAY COUNTY            0.0       57839.24            0.0   
132961          FL  CLAY COUNTY            0.0     5499407.39            0.0   
418645          FL  CLAY COUNTY            0.0       57235.46            0.0   
451133          FL  CLAY COUNTY            0.0       39279.24            0.0   
688245          FL  CLAY COUNTY            0.0        5611.32            0.0   
516626          FL  CLAY COUNTY            0.0        3193.31            0.0   
936566          FL  CLAY COUNTY            0.0      645997.97            0.0   
447164          FL  CLAY COUNTY            0.0       36914.17            0.0   
881403          FL  CLAY COUNTY        10935.0       10935.00        10935.0   
890018          FL  CLAY COUNTY            0.0      177534.72            0.0   
298166          FL  CLAY COUNTY            0.0      627738.04            0.0   
201251          FL  CLAY COUNTY            0.0        3793.74            0.0   
484359   

# Access Multiple columns

In [58]:
df[['county','statecode','point_granularity']]

county statecode  point_granularity
policyID                                            
119736      CLAY COUNTY        FL                1.0
448094      CLAY COUNTY        FL                3.0
206893      CLAY COUNTY        FL                1.0
333743      CLAY COUNTY        FL                3.0
172534      CLAY COUNTY        FL                1.0
785275      CLAY COUNTY        FL                3.0
995932      CLAY COUNTY        FL                1.0
223488      CLAY COUNTY        FL                1.0
433512      CLAY COUNTY        FL                1.0
142071      CLAY COUNTY        FL                1.0
253816      CLAY COUNTY        FL                1.0
894922      CLAY COUNTY        FL                1.0
422834      CLAY COUNTY        FL                1.0
582721      CLAY COUNTY        FL                1.0
842700      CLAY COUNTY        FL                1.0
874333      CLAY COUNTY        FL                1.0
580146      CLAY COUNTY        FL                3.0
456149      CLAY COUNTY        FL                1.0
767862      CLAY COUNTY        FL                3.0
353022      CLAY COUNTY        FL                1.0
367814      CLAY COUNTY        FL                1.0
671392      CLAY COUNTY        FL                3.0
772887      CLAY COUNTY        FL                1.0
983122      CLAY COUNTY        FL                3.0
934215      CLAY COUNTY        FL                1.0
385951      CLAY COUNTY        FL                1.0
716332      CLAY COUNTY        FL                1.0
751262      CLAY COUNTY        FL                1.0
633663      CLAY COUNTY        FL                1.0
105851      CLAY COUNTY        FL                1.0
...                 ...       ...                ...
604815    ORANGE COUNTY        FL                4.0
123581    ORANGE COUNTY        FL                1.0
830561    ORANGE COUNTY        FL                1.0
567945    ORANGE COUNTY        FL                1.0
304010    ORANGE COUNTY        FL                3.0
202453    ORANGE COUNTY        FL                1.0
920181    ORANGE COUNTY        FL                3.0
186980    ORANGE COUNTY        FL                3.0
350034    ORANGE COUNTY        FL                3.0
292859    ORANGE COUNTY        FL                1.0
286864    ORANGE COUNTY        FL                1.0
820966    ORANGE COUNTY        FL                1.0
756126    ORANGE COUNTY        FL                1.0
617621    ORANGE COUNTY        FL                1.0
950327    ORANGE COUNTY        FL                1.0
796959    ORANGE COUNTY        FL                1.0
270198    ORANGE COUNTY        FL                1.0
809530    ORANGE COUNTY        FL                1.0
474787    ORANGE COUNTY        FL                4.0
260965    ORANGE COUNTY        FL                1.0
613939    ORANGE COUNTY        FL                1.0
508003    ORANGE COUNTY        FL                1.0
638040    ORANGE COUNTY        FL                1.0
826497    ORANGE COUNTY        FL                3.0
137757    ORANGE COUNTY        FL                1.0
906943    ORANGE COUNTY        FL                1.0
758044    ORANGE COUNTY        FL                1.0
387570    ORANGE COUNTY        FL                3.0
310787    ORANGE COUNTY        FL                3.0
522718    ORANGE COUNTY        FL                NaN

[13571 rows x 3 columns]

In [59]:
df[['county']]

county
policyID               
119736      CLAY COUNTY
448094      CLAY COUNTY
206893      CLAY COUNTY
333743      CLAY COUNTY
172534      CLAY COUNTY
785275      CLAY COUNTY
995932      CLAY COUNTY
223488      CLAY COUNTY
433512      CLAY COUNTY
142071      CLAY COUNTY
253816      CLAY COUNTY
894922      CLAY COUNTY
422834      CLAY COUNTY
582721      CLAY COUNTY
842700      CLAY COUNTY
874333      CLAY COUNTY
580146      CLAY COUNTY
456149      CLAY COUNTY
767862      CLAY COUNTY
353022      CLAY COUNTY
367814      CLAY COUNTY
671392      CLAY COUNTY
772887      CLAY COUNTY
983122      CLAY COUNTY
934215      CLAY COUNTY
385951      CLAY COUNTY
716332      CLAY COUNTY
751262      CLAY COUNTY
633663      CLAY COUNTY
105851      CLAY COUNTY
...                 ...
604815    ORANGE COUNTY
123581    ORANGE COUNTY
830561    ORANGE COUNTY
567945    ORANGE COUNTY
304010    ORANGE COUNTY
202453    ORANGE COUNTY
920181    ORANGE COUNTY
186980    ORANGE COUNTY
350034    ORANGE COUNTY
292859    ORANGE COUNTY
286864    ORANGE COUNTY
820966    ORANGE COUNTY
756126    ORANGE COUNTY
617621    ORANGE COUNTY
950327    ORANGE COUNTY
796959    ORANGE COUNTY
270198    ORANGE COUNTY
809530    ORANGE COUNTY
474787    ORANGE COUNTY
260965    ORANGE COUNTY
613939    ORANGE COUNTY
508003    ORANGE COUNTY
638040    ORANGE COUNTY
826497    ORANGE COUNTY
137757    ORANGE COUNTY
906943    ORANGE COUNTY
758044    ORANGE COUNTY
387570    ORANGE COUNTY
310787    ORANGE COUNTY
522718    ORANGE COUNTY

[13571 rows x 1 columns]

df['county'] - It is a series

df[['county']] - DataFrame

Access row with index 133 as a series

In [60]:
df.iloc[133]

statecode                           FL
county                ST  JOHNS COUNTY
eq_site_limit                        0
hu_site_limit                   159042
fl_site_limit                        0
fr_site_limit                        0
tiv_2011                        159042
tiv_2012                        249985
eq_site_deductible                   0
hu_site_deductible                   0
fl_site_deductible                   0
fr_site_deductible                   0
point_latitude                 29.7766
point_longitude               -81.4717
line                       Residential
construction                      Wood
point_granularity                    1
Name: 482457, dtype: object

In [61]:
df.iloc[[133]]

statecode            county  eq_site_limit  hu_site_limit  \
policyID                                                             
482457          FL  ST  JOHNS COUNTY            0.0      159041.52   

          fl_site_limit  fr_site_limit   tiv_2011   tiv_2012  \
policyID                                                       
482457              0.0            0.0  159041.52  249984.64   

          eq_site_deductible  hu_site_deductible  fl_site_deductible  \
policyID                                                               
482457                   0.0                 0.0                 0.0   

          fr_site_deductible  point_latitude  point_longitude         line  \
policyID                                                                     
482457                     0       29.776566        -81.47171  Residential   

         construction  point_granularity  
policyID                                  
482457           Wood                1.0

In [62]:
df.iloc[133:134]

statecode            county  eq_site_limit  hu_site_limit  \
policyID                                                             
482457          FL  ST  JOHNS COUNTY            0.0      159041.52   

          fl_site_limit  fr_site_limit   tiv_2011   tiv_2012  \
policyID                                                       
482457              0.0            0.0  159041.52  249984.64   

          eq_site_deductible  hu_site_deductible  fl_site_deductible  \
policyID                                                               
482457                   0.0                 0.0                 0.0   

          fr_site_deductible  point_latitude  point_longitude         line  \
policyID                                                                     
482457                     0       29.776566        -81.47171  Residential   

         construction  point_granularity  
policyID                                  
482457           Wood                1.0

In [63]:
df.county.isnull()

policyID
119736    False
448094    False
206893    False
333743    False
172534    False
785275    False
995932    False
223488    False
433512    False
142071    False
253816    False
894922    False
422834    False
582721    False
842700    False
874333    False
580146    False
456149    False
767862    False
353022    False
367814    False
671392    False
772887    False
983122    False
934215    False
385951    False
716332    False
751262    False
633663    False
105851    False
          ...  
604815    False
123581    False
830561    False
567945    False
304010    False
202453    False
920181    False
186980    False
350034    False
292859    False
286864    False
820966    False
756126    False
617621    False
950327    False
796959    False
270198    False
809530    False
474787    False
260965    False
613939    False
508003    False
638040    False
826497    False
137757    False
906943    False
758044    False
387570    False
310787    False
522718    False
Name: county, L

In [64]:
df.county.isna()

policyID
119736    False
448094    False
206893    False
333743    False
172534    False
785275    False
995932    False
223488    False
433512    False
142071    False
253816    False
894922    False
422834    False
582721    False
842700    False
874333    False
580146    False
456149    False
767862    False
353022    False
367814    False
671392    False
772887    False
983122    False
934215    False
385951    False
716332    False
751262    False
633663    False
105851    False
          ...  
604815    False
123581    False
830561    False
567945    False
304010    False
202453    False
920181    False
186980    False
350034    False
292859    False
286864    False
820966    False
756126    False
617621    False
950327    False
796959    False
270198    False
809530    False
474787    False
260965    False
613939    False
508003    False
638040    False
826497    False
137757    False
906943    False
758044    False
387570    False
310787    False
522718    False
Name: county, L

In [65]:
sample = pd.read_csv("sample.csv")
sample

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [66]:
sample.salary.isna()

0    False
1    False
2    False
3    False
Name: salary, dtype: bool

In [67]:
sample.salary.isnull()

0    False
1    False
2    False
3    False
Name: salary, dtype: bool

In [68]:
sample.isnull()

name  salary   year
0  False   False  False
1  False   False  False
2  False   False  False
3  False   False  False

In [69]:
sample.isna()

name  salary   year
0  False   False  False
1  False   False  False
2  False   False  False
3  False   False  False

In [70]:
True+False+True # Number of true

2

In [71]:
sample.isna().sum() # default axis=0

name      0
salary    0
year      0
dtype: int64

In [72]:
sample.isna().sum(axis=1)

0    0
1    0
2    0
3    0
dtype: int64

In [73]:
sample[sample.salary.isna()]

Empty DataFrame
Columns: [name, salary, year]
Index: []

In [74]:
sample.loc[sample.salary.isna()]

Empty DataFrame
Columns: [name, salary, year]
Index: []

# To remove columns

In [75]:
# To remove a column with label salary
sample.drop("salary",axis=1)
# inplace =TRue will overwrite the original column
# sample = sample.drop("salary",axis=1)

name  year
0      Raj Acharya  2015
1      SaiTeja Ram  2012
2    Hemendra Arya  2018
3  Revanth Sundarm  2019

In [78]:

# sample.drop("salary")
# sample.drop("hike",axis=1)

In [79]:
# To remove columns with labels salary and year
sample.drop(["salary",'year'],axis=1)

name
0      Raj Acharya
1      SaiTeja Ram
2    Hemendra Arya
3  Revanth Sundarm

In [80]:
# To remove columns with indices 1 and 2
sample.drop(index = [1,2],axis=1)

name  salary  year
0      Raj Acharya   45000  2015
3  Revanth Sundarm   15000  2019

In [81]:
sample

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [82]:
help(sample.drop)

Help on method drop in module pandas.core.frame:

drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise') method of pandas.core.frame.DataFrame instance
    Drop specified labels from rows or columns.
    
    Remove rows or columns by specifying label names and corresponding
    axis, or by specifying directly index or column names. When using a
    multi-index, labels on different levels can be removed by specifying
    the level.
    
    Parameters
    ----------
    labels : single label or list-like
        Index or column labels to drop.
    axis : {0 or 'index', 1 or 'columns'}, default 0
        Whether to drop labels from the index (0 or 'index') or
        columns (1 or 'columns').
    index, columns : single label or list-like
        Alternative to specifying axis (``labels, axis=1``
        is equivalent to ``columns=labels``).
    
        .. versionadded:: 0.21.0
    level : int or level name, optional
        For MultiIndex, level

# To remove rows

In [83]:
# To remove rows with lables 0 and 1
sample.drop([0,1],axis=0)

name  salary  year
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [84]:

# To remove rows with indices 0 and 1
sample.drop(index = [0,1],axis=0)

name  salary  year
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [85]:
# To remove a single row with index 3
sample.drop(3,axis=0)

name  salary  year
0    Raj Acharya   45000  2015
1    SaiTeja Ram   78000  2012
2  Hemendra Arya   32000  2018

In [87]:
 sample.drop(index=5,axis=0)

KeyError: '[5] not found in axis'

# To remove rows with missing values

In [88]:
sample.dropna() # Removes the rows where we have missing values

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [89]:
sample.dropna(axis=1) # Removes the columns where we have missing values

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [90]:
sample.salary.dropna()

0    45000
1    78000
2    32000
3    15000
Name: salary, dtype: int64

# To fill the missing value with 0

In [91]:
sample.fillna(0)

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [92]:
sample.salary.fillna(0)
# sample.salary = sample.salary.fillna(0)

0    45000
1    78000
2    32000
3    15000
Name: salary, dtype: int64

# Fill the missing value with mean

In [93]:

m =  sample.salary.mean()
print(m)
sample.salary.fillna(m)

42500.0


0    45000
1    78000
2    32000
3    15000
Name: salary, dtype: int64

In [94]:
# Total Missing value count in the data
sample.isnull().sum().sum()

0

In [95]:
# To get misisng value count in a column
sample.salary.isnull().sum()

0

In [96]:
sample

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [97]:
sample_out = sample.copy()
sample_out['salary'] = sample_out.salary.fillna(sample_out.salary.mean())
sample_out

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [98]:
sample_out_rows = sample.dropna()
sample_out_rows

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [99]:
sample_out_cols = sample.dropna(axis=1)
sample_out_cols

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [100]:
dummy = sample.copy()
dummy

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [101]:

dummy.salary = dummy.salary.mean()
dummy

name   salary  year
0      Raj Acharya  42500.0  2015
1      SaiTeja Ram  42500.0  2012
2    Hemendra Arya  42500.0  2018
3  Revanth Sundarm  42500.0  2019

In [102]:

dummy = sample.copy()
dummy

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [103]:
dummy.salary = dummy.salary.mean
dummy

name                                             salary  year
0      Raj Acharya  <bound method Series.mean of 0    45000\n1    ...  2015
1      SaiTeja Ram  <bound method Series.mean of 0    45000\n1    ...  2012
2    Hemendra Arya  <bound method Series.mean of 0    45000\n1    ...  2018
3  Revanth Sundarm  <bound method Series.mean of 0    45000\n1    ...  2019

# pd.get_dummies

- It will create dummy columns with unique values as column names
- One hot encoded data

In [104]:
s = pd.Series(["avg","good","bad","avg","good"])
pd.get_dummies(s)

avg  bad  good
0    1    0     0
1    0    0     1
2    0    1     0
3    1    0     0
4    0    0     1

In [105]:
s.nunique()

3

In [106]:
s.unique()

array(['avg', 'good', 'bad'], dtype=object)

In [107]:
sample

name  salary  year
0      Raj Acharya   45000  2015
1      SaiTeja Ram   78000  2012
2    Hemendra Arya   32000  2018
3  Revanth Sundarm   15000  2019

In [108]:
pd.get_dummies(sample)

salary  year  name_Hemendra Arya  name_Raj Acharya  name_Revanth Sundarm  \
0   45000  2015                   0                 1                     0   
1   78000  2012                   0                 0                     0   
2   32000  2018                   1                 0                     0   
3   15000  2019                   0                 0                     1   

   name_SaiTeja Ram  
0                 0  
1                 1  
2                 0  
3                 0

In [109]:
data.dtypes==object

policyID              False
statecode              True
county                 True
eq_site_limit         False
hu_site_limit         False
fl_site_limit         False
fr_site_limit         False
tiv_2011              False
tiv_2012              False
eq_site_deductible    False
hu_site_deductible    False
fl_site_deductible    False
fr_site_deductible    False
point_latitude        False
point_longitude       False
line                   True
construction           True
point_granularity     False
dtype: bool

In [110]:
data.loc[:,data.dtypes==object].nunique()

statecode        1
county          42
line             2
construction     5
dtype: int64

In [111]:
data.loc[:,data.dtypes==object].nunique().sum()

50

In [112]:

catageroical_columns_count = (data.dtypes==object).sum()
catageroical_columns_count

4

In [113]:
# no of columns in the output
75 - 4 + 18

89

In [114]:
pd.get_dummies(data)

policyID  eq_site_limit  hu_site_limit  fl_site_limit  fr_site_limit  \
0        119736       498960.0      498960.00       498960.0       498960.0   
1        448094      1322376.3     1322376.30      1322376.3      1322376.3   
2        206893       190724.4      190724.40       190724.4       190724.4   
3        333743            0.0       79520.76            0.0            0.0   
4        172534            0.0      254281.50            0.0       254281.5   
5        785275            0.0      515035.62            0.0            0.0   
6        995932            0.0    19260000.00            0.0            0.0   
7        223488       328500.0      328500.00       328500.0       328500.0   
8        433512       315000.0      315000.00       315000.0       315000.0   
9        142071       705600.0      705600.00       705600.0       705600.0   
10       253816       831498.3      831498.30       831498.3       831498.3   
11       894922            0.0       24059.09            0.0            0.0   
12       422834            0.0       48115.94            0.0            0.0   
13       582721            0.0       28869.12            0.0            0.0   
14       842700            0.0       56135.64            0.0            0.0   
15       874333            0.0       48115.94            0.0            0.0   
16       580146            0.0       48115.94            0.0            0.0   
17       456149            0.0       80192.49            0.0            0.0   
18       767862            0.0       48115.94            0.0            0.0   
19       353022            0.0       60946.79            0.0            0.0   
20       367814            0.0       28869.12            0.0            0.0   
21       671392            0.0    13410000.00            0.0            0.0   
22       772887            0.0     1669113.93            0.0            0.0   
23       983122            0.0      179562.23            0.0            0.0   
24       934215            0.0      177744.16            0.0            0.0   
25       385951            0.0       17757.58            0.0            0.0   
26       716332            0.0      130129.87            0.0            0.0   
27       751262            0.0       42854.77            0.0            0.0   
28       633663            0.0         785.58            0.0            0.0   
29       105851            0.0      170361.91            0.0            0.0   
...         ...            ...            ...            ...            ...   
13541    604815       497718.0      497718.00       497718.0       497718.0   
13542    123581      1083600.0     1083600.00      1083600.0      1083600.0   
13543    830561       418230.0      418230.00       418230.0       418230.0   
13544    567945            0.0      277276.92            0.0            0.0   
13545    304010            0.0    14130000.00            0.0     14130000.0   
13546    202453            0.0    19260000.00            0.0     19260000.0   
13547    920181            0.0    19620000.00            0.0     19620000.0   
13548    186980            0.0     8432564.40            0.0      8432564.4   
13549    350034            0.0     2858963.40            0.0      2858963.4   
13550    292859            0.0    19620000.00            0.0     19620000.0   
13551    286864            0.0      681474.42            0.0            0.0   
13552    820966            0.0     3941825.58            0.0            0.0   
13553    756126            0.0      190479.96            0.0            0.0   
13554    617621            0.0      563467.50            0.0       563467.5   
13555    950327            0.0     1905773.72            0.0            0.0   
13556    796959            0.0     1391883.85            0.0            0.0   
13557    270198            0.0     4534866.90            0.0      4533966.9   
13558    809530            0.0     1656192.60            0.0      1656192.6   
13559    474787     10530000.0    10530000.00     1053000

In [115]:
pd.get_dummies(data).shape[1]

64